### ISSUE TITLE
Support a --sphinx-doctest option

### ISSUE THREAD

**Issue by jab** [pytest-dev/pytest#4274]
*(Split off #2514)*

I love the convenience of being able to invoke a single, well-configured `pytest` command to run all of the many tests in my project, including the example code in my Sphinx docs. (I even contributed #1255 to improve support for this:)

But a couple of the missing [sphinx doctest features](http://www.sphinx-doc.org/en/master/usage/extensions/doctest.html) are becoming show-stoppers. (Specifically pyversion, skipif, and testsetup. Happy to explain why if that'd be helpful. I think my use case is pretty typical.)

But given that Sphinx already supports all of these with its own `make doctest` command, I'd be plenty happy if `pytest` would just let me pass it some `--sphinx-doctest=path/to/docs` option, to have it shell out to Sphinx's own `make doctest` command as part of what happens when I run `pytest`. As long as pytest made sure to exit nonzero if `make doctest` exited nonzero, and had no other integration, I think it'd still be a win over what pytest currently supports via `--doctest-modules`.

Such a `--sphinx-doctest` option seems simple to implement, and would not impose much maintenance burden. It's just a matter of gluing a command line option to a `subprocess.run()` call.

Would you consider accepting a PR that implemented this?

(I am aware of the https://github.com/thisch/pytest-sphinx plugin, but it is still missing some of these features too, and it could be worth avoiding an extra plugin dependency since Sphinx already does what we need.)

Thanks for your consideration, and for the great work on pytest!

**Comment by thisch** [comment]
If you shell out you miss a nice feature like measuring of the code coverage and you don't see the run sphinx tests in `junit.xml` (if you use that). Note that `sphinx-build -b doctest` may run tests that were executed by pytest a second time, e.g., if you use the syntax that is supported by the doctest module from the stdlib in rst files.

**Comment by nicoddemus** [comment]
Hi @jab,

Using `make doctest` is not a good approach as we would then depend on an external executable, and it wouldn't work on Windows.

> Such a --sphinx-doctest option seems simple to implement, and would not impose much maintenance burden. It's just a matter of gluing a command line option to a subprocess.run() call.

I understand this might look like a simple implementation, but in my experience this kind of quick workaround is the source of many bug reports down the road.

> I am aware of the thisch/pytest-sphinx plugin, but it is still missing some of these features too, and it could be worth avoiding an extra plugin dependency since Sphinx already does what we need.

I think it would be worthwhile to add the functionality you miss to `pytest-sphinx` then. @thisch is very approachable and I'm sure he would welcome improvements to the plugin.

**Comment by jab** [comment]
Thanks @thisch and @nicoddemus. I spent a few hours trying to hack something into `pytest-sphinx` for this but didn't get anywhere. It may just be that I've never worked on any doctest, sphinx, or pytest plugin code individually, and so with the combination of all three, it's especially hard to parachute into the `pytest-sphinx` code and cleanly add these features. But from the TODO messages in that code (e.g. "TODO: subclass doctest.DocTestParser instead?", "CLEANUP: use the sphinx directive parser from the sphinx project", etc.), as well as the lack of progress after a few hours, I think it might also be that the `pytest-sphinx` code needs to be refactored more substantially before these features can be added cleanly by an external contributor.

(It's clear to me what it's currently doing, and I was able to extend it to get a hold of pairs of `testsetup` and `doctest` sections, and then add the `doctest` content to the parsed `examples`, but it's just not at all clear how/where to feed the `testsetup` content such that it's executed at the right time and the side effects happen in the right namespace. It's clear to me how https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/doctest.py is doing this, but `pytest-sphinx` is doing things completely differently.)

I'd love to see this added to `pytest-sphinx`, but I'm unfortunately out of time to keep working on it, and to be honest already spent more time than I should have on something that was already working well enough for me via sphinx's `make doctest`.

In any case, I'll close this issue since it doesn't seem like it will be considered further, but thanks again for your consideration.
